src/ocean/util/container/btree/BTreeRange.d

```diff
@@ -0,0 +1,355 @@
+
+*******************************************************************************/
+
+public struct BTreeRange(BTree)
+{
+    /// Root node of a tree
+    private BTree.Implementation* tree;
+
+    /// "Version" of the tree at the iteration time.
+    /// Used to check if the iterator is invalid
+    private ulong tree_version;
+
+    /// Pair of the node/index for which we've paused iteration and pushed
+    /// to stack
+    private struct IterationItem
+    {
+        BTree.Implementation.BTreeNode* node;
+        size_t index;
+    }
+
+    /// Stack of the nodes that we've postponed iteration over
+    private IterationItem[]* stack;
+
+    /// Current item we're on
+    private IterationItem item;
+
+    /// Current element over which we're iterating
+    private BTree.ElementType* current_element;
+
+    /***************************************************************************
+
+        Returns:
+            the current element in the range.
+
+        Throws:
+           RangeInvalidatedException if the underlying tree has changed
+
+    ***************************************************************************/
+
+    public BTree.ElementType front ()
+    {
+        this.ensureValid();
+        return *this.current_element;
+    }
+
+    /***************************************************************************
+
+        Pops the next element from the tree.
+
+        Throws:
+           RangeInvalidatedException if the underlying tree has changed
+
+    ***************************************************************************/
+
+    public void popFront ()
+    {
+        this.ensureValid();
+        this.iterationStep();
+    }
+
+    /***************************************************************************
+
+        Returns:
+            true if there are no more elements to iterate over
+
+        Throws:
+           RangeInvalidatedException if the underlying tree has changed
+
+    ***************************************************************************/
+
+    public bool empty ()
+    {
+        this.ensureValid();
+        return this.tree.root is null ||
+            (this.item.node is null && this.stack.length == 0);
+    }
+
+    /***************************************************************************
+
+        Returns:
+            true if the underlying tree has not changed and it's safe
+            to use front/popFront/empty methods.
+
+    ***************************************************************************/
+
+    public bool isValid ()
+    {
+        return tree_version == this.tree.content_version;
+    }
+
+    /***************************************************************************
+
+        Ensures that a tree has not changed since the creation of this range.
+
+        Throws:
+           RangeInvalidatedException if the underlying tree has changed
+
+    ***************************************************************************/
+
+    private void ensureValid ()
```

Comment:
This is now tricky. Because it doesn't _enforce_ anything, it just ensures that the range is valid before accessing it.